feat: internal provider state, new client events (#241)
open-feature/spec: specification/sections/05-events.md

> The `provider` **MAY** define a mechanism for signaling the occurrence of one of a set of events, including `PROVIDER_READY`, `PROVIDER_ERROR`, `PROVIDER_CONFIGURATION_CHANGED` and `PROVIDER_STALE`, with a `provider event details` payload. 

> Providers cannot emit `PROVIDER_CONTEXT_CHANGED` or `PROVIDER_CONTEXT_PENDING` event. These are emitted only by the SDK during context reconciliation.

Reviewer: ```suggestion
Providers cannot emit `PROVIDER_CONTEXT_CHANGED` or `PROVIDER_CONTEXT_PENDING` event.
These are emitted only by the SDK during context reconciliation.
```

If available, native event-emitter or observable/observer language constructs can be used.

When a provider is unable to evaluate flags (perhaps due to loss of connection with a remote service) the provider can signal this by emitting a `PROVIDER_ERROR` event.

> If the error state is irrecoverable, the `PROVIDER_STATE_FATAL` error code can be used.

Reviewer: Is it possible to set `PROVIDER_STATE_FATAL` during initialization? This may happen if the access key is invalid.

Author: Yes, that's one of the main purposes. I can mention it in the init section as well.

Author: I've done this.

> The `provider` **MAY** define a mechanism for signaling the occurrence of one of a set of events, including `PROVIDER_READY`, `PROVIDER_ERROR`, `PROVIDER_CONFIGURATION_CHANGED` and `PROVIDER_STALE`, with a `provider event details` payload. 

> Providers cannot emit `PROVIDER_CONTEXT_CHANGED` or `RECONCILING` event.

Reviewer: ```suggestion
Providers cannot emit `PROVIDER_CONTEXT_CHANGED` or `PROVIDER_RECONCILING` event.
```

At least that's what the event is called in the types doc.

Author: Thanks!

These are emitted only by the SDK during context reconciliation.

If available, native event-emitter or observable/observer language constructs can be used.

When a provider is unable to evaluate flags (perhaps due to loss of connection with a remote service) the provider can signal this by emitting a `PROVIDER_ERROR` event.

> If the error state is irrecoverable, the `PROVIDER_FATAL` error code can be used.
> When it recovers, it can emit a `PROVIDER_READY` event.

Reviewer: ```suggestion
When it recovers, it can emit a `PROVIDER_READY` event.
If the error state is irrecoverable, the `PROVIDER_FATAL` error code can be used.
```

I would swap these lines because a `PROVIDER_FATAL` error code isn't recoverable but a `PROVIDER_ERROR` is.

### Event handlers and context reconciliation

Providers built to conform to the static context paradigm feature two additional events: `PROVIDER_CONTEXT_CHANGE_PENDING` and `PROVIDER_CONTEXT_CHANGED`.

> While the provider is reconciling it's internal state (the `on context changed` function is running and not yet terminated) the SDK emits `PROVIDER_CONTEXT_CHANGE_PENDING` and transitions the provider into state `CONTEXT_PENDING`.

Reviewer: ```suggestion
While the provider is reconciling its internal state (the `on context changed` function is running and not yet terminated), the SDK emits `PROVIDER_CONTEXT_CHANGE_PENDING` and transitions the provider into state `CONTEXT_PENDING`.
```

This can be particularly useful for displaying loading indicators while the [evaluation context](./03-evaluation-context.md) is being reconciled.

> If the `on context changed` function terminates normally, the SDK emits (`PROVIDER_CONTEXT_CHANGED`) and transitions the provider into the `READY` state, otherwise it emits `PROVIDER_ERROR` and transitions the provider into `ERROR` state.

Reviewer 2: How do we handle cases where `setEvaluationContext()` is called a second time by the application before the first call has not finished yet? Should the first `onContextChange()` call terminate normally or abnormally, and what are the implications with the Events emitted and the State changes?

Author: That's a great question.

My first thoughts are that we shouldn't terminate abnormally, and should emit events as normal. We may want to consider adding requirements on providers to make sure the function is reentrant, so at least there's no accidental errors if this happens.

We could also guarantee this doesn't happen by queueing up context-changes, but I think that's putting quite a burden on the SDK for a bit of an edge case. Seeing that `setContext` is an async operation in most implementations, I think authors inherently expect that multiple calls might not finish in the order they are started, and so should do what they can to avoid such race conditions.

A middle ground might be to never emit `PROVIDER_CONTEXT_CHANGED` or change the state from `CONTEXT_PENDING` while another reconciliation is in progress, which seems reasonable, so at least if the function is called twice, we wouldn't pop in and out of the pending state.

Reviewer 3: Is this really an edge-case? I would imagine apps call setContext from event handlers (say a dark-mode toggle) which would quickly lead to overlapping reconciliation cycles, and the spec should be clear on how to handle these.

Also, from an experimentation perspective, it's adamant that the state remain in RECONCILING until the last initiated cycle has ended. If the state transitions to READY before the last cycle has ended, flags will return stale values (for example values for another user). If the state transitions to ERROR before the last cycle has ended, flags will return default values. Both of these behaviours will taint the user experience and the experiment.

Author: > Is this really an edge-case? I would imagine apps call setContext from event handlers (say a dark-mode toggle) which would quickly lead to overlapping reconciliation cycles, and the spec should be clear on how to handle these.
> 
> Also, from an experimentation perspective, it's adamant that the state remain in RECONCILING until the last initiated cycle has ended. If the state transitions to READY before the last cycle has ended, flags will return stale values (for example values for another user). If the state transitions to ERROR before the last cycle has ended, flags will return default values. Both of these behaviours will taint the user experience and the experiment.

@andreas-karlsson you make some compelling points.

We could keep things vague and just say SDKs must "handle" this in some way... but I'm thinking it's better to prescribe a pattern to handle it.

Do you think if we spec'd something like what I described [here](https://github.com/open-feature/spec/pull/241#discussion_r1479925311) (specifically the quoted section below) it would be sufficient?

> A middle ground might be to never emit PROVIDER_CONTEXT_CHANGED or change the state from CONTEXT_PENDING while another reconciliation is in progress, which seems reasonable, so at least if the function is called twice, we wouldn't pop in and out of the pending state.

cc @fabriziodemaria

Reviewer 3: Thanks, that would be great.

Author: @andreas-karlsson @fabriziodemaria I've pushed a commit to address this specific issue: https://github.com/open-feature/spec/pull/241/commits/ad3b33768f9de746b9c766164b186b62d7dcaddc

### Event handlers and context reconciliation

Providers built to conform to the static context paradigm feature two additional events: `PROVIDER_CONTEXT_CHANGE_PENDING` and `PROVIDER_CONTEXT_CHANGED`.
While the provider is reconciling its internal state (the `on context changed` function is running and not yet terminated), the SDK emits `PROVIDER_CONTEXT_CHANGE_PENDING` and transitions the provider into state `CONTEXT_PENDING`.
This can be particularly useful for displaying loading indicators while the [evaluation context](./03-evaluation-context.md) is being reconciled.

If the `on context changed` function terminates normally, the SDK emits (`PROVIDER_CONTEXT_CHANGED`) and transitions the provider into the `READY` state, otherwise it emits `PROVIDER_ERROR` and transitions the provider into `ERROR` state.

> The `PROVIDER_CONTEXT_CHANGED` is used to signal that the associated context has been changed, and flags should be re-evaluated.

Author: > If we change to `RECONCILED` we should also change to something like `PROVIDER_RECONCILED`.

We could, and I consider that as well, but this would be a breaking change to any of the implementing client providers (including web). I felt that this wasn't worth it, but I could be convinced otherwise.

> The `PROVIDER_CONTEXT_CHANGED` is used to signal that the associated context has been changed, and flags should be re-evaluated.

Reviewer 4: Ah no sorry I did not consider that this already existed.
I would also leave it then.

### Event handlers and context reconciliation

> Providers built to conform to the static context paradigm feature two additional events: `PROVIDER_CONTEXT_CHANGE_PENDING` and `PROVIDER_CONTEXT_CHANGED`.

Reviewer 2: ```suggestion
Providers built to conform to the static context paradigm feature two additional events: `PROVIDER_RECONCILING` and `PROVIDER_CONTEXT_CHANGED`.
```
This is the naming used in the "types" table. I see a couple more instances of `PROVIDER_CONTEXT_CHANGE_PENDING` throughout the PR, which I think we need to update

Author: Nice catch. I will make sure these are all resolved.

##### Conditional Requirement 5.3.4.1

> While the provider's `on context changed` function is executing, associated `PROVIDER_CONTEXT_PENDING` handlers **MUST** run.

The implementation must run any `PROVIDER_CONTEXT_PENDING` handlers associated with the provider while the provider is reconciling its state.

> In languages with asynchronous semantics, the emission of this event can be skipped if the `on context changed` function of the provider in question executes synchronously for a given provider, no other operations can take place while it runs.

Reviewer: Why are events skipped in `on context changed` is synchronous?

Author: If it's synchronous, there's no point emitting a pending event. This would be for providers that don't need to do any async context reconciliation, they just do some local computation. In languages like JS, while that's happening, nothing else can happen, so there' no point in emitting the event. Put another way, the context was reconciled in "real time", so it was never "pending" to the outside world - we'd just skip right to "changed".

In JS, this would look like the `on context change` function of the provider didn't return a promise but just did some synchronous work (returned void). Events would be useless or even potentially cause a disruptive UI flash.

I just wanted to provide this escape hatch here. I'm hesitant to go into more detail in the spec, but I wanted to give implementations the ability to skip this emission because I think it makes sense.

#### Requirement 5.3.5

> If the provider emits an event, the value of the client's `provider status` **MUST** be updated accordingly.

Some providers may emit events spontaneously, based on changes in their internal state (connections, caches, etc).
The SDK must update it's internal representation of the provider's state accordingly:

| Event                            | Associated Status                                                  |
| -------------------------------- | ------------------------------------------------------------------ |
| `PROVIDER_READY`                 | `READY`                                                            |
| `PROVIDER_STALE`                 | `STALE`                                                            |
| `PROVIDER_ERROR`                 | `ERROR`                                                            |
| `PROVIDER_CONFIGURATION_CHANGED` | N/A (provider remains in state `READY`)                            |
| `PROVIDER_CONTEXT_CHANGED`       | N/A (only emitted by SDK during context reconciliation) |

> | `PROVIDER_CONTEXT_PENDING`       | N/A (only emitted by SDK during context reconciliation) |

Reviewer: Please format the table.

The `PROVIDER_ERROR` is not emitted from the provider itself; the SDK implementation must run the `PROVIDER_ERROR` handlers if the `on context changed` throws or otherwise signals an error.

> see: [provider event types](../types.md#provider-events), [provider events](#51-provider-events), context, [provider context reconciliation](02-providers.md#26-provider-context-reconciliation)

Reviewer: Should `context` be a link?

Author: I think this is actually a typo, I think just linking to `provider context reconciliation` is enough.

#### Requirement 5.3.5

> If the provider emits an event, the value of the client's `provider status` **MUST** be updated accordingly.

Some providers may emit events spontaneously, based on changes in their internal state (connections, caches, etc).
The SDK must update it's internal representation of the provider's state accordingly:

| Event                            | Associated Status                                       |
| -------------------------------- | ------------------------------------------------------- |
| `PROVIDER_READY`                 | `READY`                                                 |
| `PROVIDER_STALE`                 | `STALE`                                                 |
| `PROVIDER_ERROR`                 | `ERROR`                                                 |

> | `PROVIDER_CONFIGURATION_CHANGED` | N/A (provider remains in state `READY`)                 |

Reviewer: ```suggestion
| `PROVIDER_CONFIGURATION_CHANGED` | N/A (provider remains in its current state)                 |
```

I think this address the non-ready states like `PROVIDER_READY` and `PROVIDER_STALE`.

| `PROVIDER_CONTEXT_CHANGED`       | N/A (only emitted by SDK during context reconciliation) |
| `PROVIDER_RECONCILING`           | N/A (only emitted by SDK during context reconciliation) |

> see [provider lifecycle management](01-flag-evaluation.md#17-provider-lifecycle-management)

Reviewer: ```suggestion
see: [provider lifecycle management](01-flag-evaluation.md#17-provider-lifecycle-management)
```
